src/client/finAuthFetch.ts

```diff
@@ -0,0 +1,144 @@
+import { finCsrfFetch } from './finCsrfFetch'
+
+/**
+ * An enhanced fetch API for a pluggable authentication mechanism
+ *
+ * Transparently handles pre-authentication, authentication fault detection, authentication, and
+ * will replay the requested resource upon successful authentication.
+ *
+ * This fetch function will utilize the finCsrfFetch function internally to execute the request
+ *
+ * Example
+ * ```typescript
+ *
+ * const result = finAuthFetch(request, {
+ * 		authenticator: {
+ * 			isAuthenticated: (response: Response) => response.status !== 401,
+ *			preauthenticate: async (request: RequestInfo) => new Request(request, {headers: {auth_header: '12345'}}),
+ *			authenticate: (response: Response) => return execute_authentication(username, password),
+ *			maxTries: 3
+ * 	}
+ * })
+ * ```
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+ *
+ * @param resource The resource to request.
+ * @param options Optional object containing custom settings and the authenticator.
+ * @returns A promise that resolves to a response object.
+ * @throws Exception on failed authentication
+ */
+export async function finAuthFetch(
+	resource: RequestInfo,
+	options?: RequestInit | RequestInitAuth
+): Promise<Response> {
+	let resp: Response | undefined
+
+	if (options && isAuthInit(options)) {
+		const authenticator = options.authenticator as RequestAuthenticator
+
+		// Pre authenticate if suppprted by authenticator
+		if (authenticator.preauthenticate) {
+			resource = await authenticator.preauthenticate(resource)
+		}
+
+		// Pipe request to the csrf fetch function
+		resp = await finCsrfFetch(resource, options)
```

Comment:
I don't think we should have a hard coded dependency on `finCsrfFetch`. One should be able to pass in an alternative fetch implementation. Then everything can be plugged together into a chain for use.

That being said, if a fetch isn't specified in the arguments then it can default back to `finCsrfFetch`.